[FLINK-28357][datastream] Disallow null elements in StreamNode#typeSerializersIn

pnowojski opened the pull request:
Otherwise tasks can not correctly determine number of inputs. This was causing an
issue where restored as finished OneInputStreamTask was waiting for MAX_WATERMARK
from two inputs, where the second one was null.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / no ****/ don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

MartijnVisser commented:
@flinkbot run azure

zentol commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamNode.java:
```diff
@@ -265,17 +266,19 @@ public void setOperatorDescription(String operatorDescription) {
+                        .toArray(TypeSerializer<?>[]::new);
     }
 
     public TypeSerializer<?>[] getTypeSerializersIn() {
```
Does the code calling this method make no assumptions about the indices?
Let's say the original list of serializers contained 2 elements, with the first one being nul
l.
How does the code accessing this array figure out that what is now at index 0 is what it requires?

pnowojski replied:
It must be kept in sync with `StreamNode#getInEdges`. Each edge has it's own input serializer. It can never happen that first element is null, second is not null. Sources have zero inputs, one input task has single input, two input tasks have two inputs, multi input tasks have N inputs.

zentol replied:
Can this change result in the array being empty? If so, can this conflict with `StreamGraph#setSerializersFrom`?

pnowojski replied:
`StreamGraph#setSerializersFrom` was not used so I removed it in this PR :)

Yes the result can be empty for source tasks.

pnowojski commented:
This code kind of always assumed serializers can be null. Both input and output. Originally, before we introduced multiple input operators, it kind of held together, as the `StreamNode` had three fields: `typeSerializerIn1`, `typeSerializerIn2` and `typeSerializerOut`. Either of those could be null. When introducing multiple inputs, I squashed the `typeSerializerInX` fields into an array, unfortunately also preserving nulls. 

I've looked into disallowing nulls into those setters, but that made the caller code quite ugly, with lots of ifs. And those nulls are used in a quite a lot of places (check `StreamGraph#setSerializers(...)` calls)

zentol commented on flink-tests/src/test/java/org/apache/flink/test/streaming/api/datastream/FinishedSourcesWatermarkITCase.java:
```diff
@@ -0,0 +1,131 @@
+                        throw new RuntimeException("Force recovery");
+                    }
+                    if (lastSuccessfulCheckpointId > 10 && expectedWatermark > watermark) {
+                        expectedWatermark = watermark;
```
may be a matter of taste, but I'd consider a `break` here to be simpler (+ reducing the while condition to `isRunning()`).

zentol commented on flink-tests/src/test/java/org/apache/flink/test/streaming/api/datastream/FinishedSourcesWatermarkITCase.java:
```diff
@@ -0,0 +1,131 @@
+                    if (lastSuccessfulCheckpointId == 5) {
+                        throw new RuntimeException("Force recovery");
+                    }
+                    if (lastSuccessfulCheckpointId > 10 && expectedWatermark > watermark) {
```
```suggestion
                    if (lastSuccessfulCheckpointId == 10) {
```
The latter condition will usually only evaluate to true the first time it is called (excluding cases where `DateTime.now()` goes back in time).

pnowojski replied:
Re this and above. This is kind of the point here. We are waiting until 10th checkpoint. On the 10th checkpoint, we are recoding what was the last emitted watermark, and remembering it to be the `expectedWatermark`. Then we are still running unknown number of iterations, until the downstream sink finally receives this `expectedWatermark`. Hence we can not get rid of `expectedWatermark > ACTUAL_DOWNSTREAM_WATERMARK.get()`.

I will try to rephrase the code a bit, to make it more explicit/clear.

zentol commented on flink-tests/src/test/java/org/apache/flink/test/streaming/api/datastream/FinishedSourcesWatermarkITCase.java:
```diff
@@ -0,0 +1,131 @@
+        @Override
+        public void run(SourceContext<String> sourceContext) throws Exception {
+            long expectedWatermark = Watermark.MAX_WATERMARK.getTimestamp();
+            long watermark = DateTime.now().getMillis();
```
Do we have a requirement that emitted watermarks are monotonous?

Doesn't `DateTime.now().getMillis()` by default just call `System.currentTimeMillis` anyway (so couldn't we use that directly)?

zentol commented on flink-tests/src/test/java/org/apache/flink/test/streaming/api/datastream/FinishedSourcesWatermarkITCase.java:
```diff
@@ -0,0 +1,131 @@
+            long watermark = DateTime.now().getMillis();
+            sourceContext.emitWatermark(new Watermark(watermark));
+
+            while (isRunning && expectedWatermark > ACTUAL_DOWNSTREAM_WATERMARK.get()) {
```
```suggestion
            while (isRunning) {
```
I'm wondering whether this conditions does more harm than good. AFAICT it is always false after CP 10, except if a max watermark hits the sink (but if we are concerned about that case then maybe we should check that explicitly).

zentol reviewed (changes requested):
While I like that the test is very close to what the user reported I found it to be a bit difficult to understand what exactly we are asserting and why the conditions in the long-running source are set up the way they are.
